RationaleMCP/0035/SpecificationText.md

```diff
@@ -0,0 +1,82 @@
+
+The files to support translation must be provided by the developer of the library. They must be stored in a subdirectory of the Resources directory of the library with the name `Language`.
+
+Two kind of files have to be provided:
+* a template file `<LibraryName>.pot`. 
+It contains all necessary information to translate all descriptions, but no translations. The pattern `<LibraryName>` denotes the toplevel class name of the library.
+* one file for each supported language with the name `<LibraryName>.<language>.po`. This file is a copy of the associated template file, but extended with the translations in the specified language. The pattern `<language>` stands for the ISO 639-1 language code, e.g. de or en.
+
+The files consist of a header and a body. All text strings are in double quotes and encoded with UTF-8 characters. Comments start with an `#` and are continued until the end of line. Spaces outside strings are ignored and used as separators. The detailed format of these files is described in [GNU gettext](https://www.gnu.org/software/gettext/manual/gettext.pdf).
+
+The header is marked with an empty msgid entry and looks like this:
+```
+# SOME DESCRIPTIVE TITLE.
+# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
+# This file is distributed under the same license as the PACKAGE package.
+# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
+#
+#, fuzzy
+msgid ""
+msgstr ""
+"Project-Id-Version: PACKAGE VERSION\n"
+"Report-Msgid-Bugs-To: \n"
+"POT-Creation-Date: 2019-03-15 10:52+0100\n"
+"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
+"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
+"Language-Team: LANGUAGE <LL@li.org>\n"
+"Language: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+```
+All general terms in the header should be replaced by specific information.
+
+In the body, directly following the header, for each descriptive text there is an entry of the form
+```
+#: <file name>:<line number>
+msgctxt "<context identifier>"
+msgid "<descriptive text>"
+msgstr "<translation>"
+```
+Only the keywords `msgctxt`, `msgid` and `msgstr` are used.
+
+At first there can be an optional comment describing the location (file name and line number) of the text to translate. Multiple occurences of the same string can be listed here, separated by space.
+
+Then, the `<context identifier>` behind the keyword `msgctxt` is the full name of the Modelica class (e.g. `Modelica.Blocks.Math.Sin` ) where the text appears in. Short class definitions do not appear here. Texts in such classes belong to the enclosing full class definition.
+
+After the `msgid` keyword the text string which shall be translated follows. It should contain the original string from the Modelica text representation. 
+Since in Modelica control sequences also start with a backslash and another backslash is used to use sequences literally or to hide double quotes, no change is required here. 

```

Comment:
I am trying to understand the relation between what should be specified specifically for the proposed used in Modelica library translations, and what should be explained just as examples of how the PO format can be utilized.

Shouldn't the emphasis here be on the actual string data content, and not which of all the available PO file mechanisms that are used to encode it in the file?

In that case, shouldn't it be sufficient to say that what goes into the `msgid` is the value of the Modelica string literal?  This would have the consequence that there is no distinction between a literal newline and `\n` in the original Modelica string literal; they both become a newline in the value of the string, and then it is up to the PO file editor to choose a quoting mechanism.

Consequently, the Modelica string literal `"a\\nb"` is evaluated to a string of length three, and one of the many alternatives for expressing this in a PO file would be this:
```
#, c-format
msgid "a\\nb"
```

From a translation point of view, I only see benefits of not making a distinction between a literal newline and `\n` in the original Modelica string literals.

Reply:
Sure, it could be sufficient to define msgid as string containinng the string literal from Modelica text. But multiline strings have to be converted to multiple strings ending with `\n`, see below. Whereby a `\n` in the middle of the string remain unchanged.

Btw.`"a\\nb"` ist a string of length four.

Reply:
Sorry, I don't know what happened to the escapes in my comment above.  I'll try to edit the comment, changing `\\n` → `\n`.

Reply:
> Sure, it could be sufficient to define msgid as string containinng the string literal from Modelica text. But multiline strings have to be converted to multiple strings ending with `\n`, see below. Whereby a `\n` in the middle of the string remain unchanged.

The point I am trying to make is that the Modelica string _literal_ should be evaluated according to Modelica string literal semantics, resulting in a string _value_.  Then, this string _value_ is what gets encoded in the gettext files using one of the available format string alternatives.  I suppose that the different format string alternatives are rather similar when it comes to escape sequences, so the Modelica string literal `"a\nb"` is likely to actually look the same in most or all format string variants.

What I don't see clearly is what the choice of format string would mean for placeholders in Modelica strings, both today and in the future.  At the moment, Modelica has placeholders described for `Text` and `Figure` annotations:
- https://specification.modelica.org/master/annotations.html#text
- https://specification.modelica.org/master/annotations.html#variable-replacements

For a Modelica string such as `"Dimensions: %{width} by %{height}"`, what difference would the choice of gettext format string make?  If gettext's `c-format` is used, would it be a problem that the string as given would be a malformed C format string?

Reply:
Modelica strings are not evaluated, they are used as they are, with all patterns unchanged. The translation has to reproduce the patterns to get the same information at the end. gettexxt itself does not know anything about the sematics of patterns, they are seen as plain text there.

Reply:
And note that gettext's c-format handling is optional in the translation tools, so I don't see any problem with it here.
This is contrast to e.g., QT's translation-handling that always seems to be active.

Reply:
I guess we can consider this conversation resolved by now?